decoders/connector/milesight/em400-udl-nb-iot/v1.0.0/payload.ts

```diff
@@ -0,0 +1,231 @@
+import { Data, DataToSend } from "@tago-io/sdk/lib/types";
+
+declare let payload: DataToSend[];
+
+/**
+ * Payload Decoder for Milesight IoT EM400-UDL
+ *
+ * @param {Object} payload - The payload object containing the data to be parsed.
+ * @param {string} group - The group identifier for the data.
+ * @param {number} port - The port number associated with the payload.
+ * @param {string} receivedTime - The time the payload was received.
+ * @returns {Array} An array of data objects extracted from the payload.
+ */
+function decodePayload(bytes: number[]) {
+  const buffer = new Buffer(bytes);
+
+  if (bytes.length < 2) {
+    throw new Error("Invalid payload size");
+  }
+
+  const payload: { [key: string]: any } = {};
+  payload.startFlag = buffer.readUInt8();
+  payload.id = buffer.readUInt16BE();
+  payload.length = buffer.readUInt16BE();
+  payload.flag = buffer.readUInt8();
+  payload.frameCnt = buffer.readUInt16BE();
+  payload.protocolVersion = buffer.readUInt8();
+  payload.firmwareVersion = buffer.readAscii(4);
+  payload.hardwareVersion = buffer.readAscii(4);
+  payload.sn = buffer.readAscii(16);
+  payload.imei = buffer.readAscii(15);
+  payload.imsi = buffer.readAscii(15);
+  payload.iccid = buffer.readAscii(20);
+  payload.csq = buffer.readUInt8();
+  payload.data_length = buffer.readUInt16BE();
+  payload.data = decodeSensorData(buffer.slice(payload.data_length));
+
+  return payload;
+}
+
+function decodeSensorData(bytes) {
+  const history: { [key: string]: string | number | boolean }[] = [];
+
+  let lastid = 0;
+  let decoded: { [key: string]: string | number | boolean } = {};
+  const buffer = new Buffer(bytes);
+  while (buffer.remaining() > 0) {
+    const channel_id = buffer.readUInt8();
+    const channel_type = buffer.readUInt8();
+
+    // check if the channel id is continuous
+    if (lastid - (channel_id & 0x0f) >= 0) {
+      history.push(decoded);
+      decoded = {};
+      lastid = 0;
+    }
+    lastid = channel_id & 0x0f;
+
+    // BATTERY
+    if (channel_id === 0x01 && channel_type === 0x75) {
+      decoded.battery = buffer.readUInt8();
+    }
+    // TEMPERATURE
+    else if (channel_id === 0x03 && channel_type === 0x67) {
+      decoded.temperature = buffer.readUInt16LE() / 10;
+    }
+    // DISTANCE
+    else if (channel_id === 0x04 && channel_type === 0x82) {
+      decoded.distance = buffer.readUInt16LE();
+    }
+    // POSITION
+    else if (channel_id === 0x05 && channel_type === 0x00) {
+      decoded.position = buffer.readUInt8() === 0 ? "normal" : "tilt";
+    }
+    // LOCATION
+    else if (channel_id === 0x06 && channel_type === 0x88) {
+      decoded.latitude = buffer.readInt32LE() / 1000000;

```

Comment:
When handling location data you must create the "location" variable. It must follow the following data structure example:
 `{
    "variable": "location",
    "value": "-23.550556,-46.633146",
    "location": {
      "lat": -23.550556,
      "lng": -46.633146
    }
  }`